alg/los.cpp

```diff
@@ -0,0 +1,232 @@
+    {
+        return z > terrainHeight;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+/************************************************************************/
+/*                        GDALIsLineOfSightVisible()                    */
+/************************************************************************/
+
+/**
+ * Check Line of Sight between two points.
+ * Both input coordinates must be within the raster coordinate bounds.
+ *
+ * This algorithm will check line of sight using a Bresenham algorithm.
+ * https://www.researchgate.net/publication/2411280_Efficient_Line-of-Sight_Algorithms_for_Real_Terrain_Data
+ * Line of sight is computed in raster coordinate space, and thus may not be appropriate.
+ * For example, datasets referenced against geographic coordinate at high latitudes may have issues.
+ *
+ * @param hBand The band to read the DEM data from. This must NOT be null.
+ * 
+ * @param xA The X location (raster column) of the first point to check on the raster.
+ *
+ * @param yA The Y location (raster row) of the first point to check on the raster.
+ * 
+ * @param zA The Z location (height) of the first point to check.
+ * 
+ * @param xB The X location (raster column) of the second point to check on the raster.
+ *
+ * @param yB The Y location (raster row) of the second point to check on the raster.
+ * 
+ * @param zB The Z location (height) of the second point to check.
+ * 
+ * @param papszOptions Options for the line of sight algorithm (currently ignored).
+ * 
+ * @return True if the two points are within Line of Sight.
+ *
+ * @since GDAL 3.9
+ */
+
+bool GDALIsLineOfSightVisible(const GDALRasterBandH hBand, const int xA,
+                              const int yA, const double zA, const int xB,
+                              const int yB, const double zB,
+                              CPL_UNUSED CSLConstList papszOptions)
+{
+    VALIDATE_POINTER1(hBand, "GDALIsLineOfSightVisible", false);
+
+    // Perform a preliminary check of the start and end points.
+    if (!IsAboveTerrain(hBand, xA, yA, zA))
+    {
+        return false;
+    }
+    if (!IsAboveTerrain(hBand, xB, yB, zB))
+    {
+        return false;
+    }
+
+    // If both X and Y are the same, no further checks are needed.
+    if (xA == xB && yA == yB)
+    {
+        return true;
+    }
+
+    // TODO if both X's or Y's are the same, it could be optimized for vertical/horizontal lines.
+
+    // Use an interpolated Z height with 2D bresenham for the remaining cases.
+
+    // Lambda for Linear interpolate like C++20 std::lerp.
+    auto lerp = [](const int a, const int b, const float t)
+    { return a + t * (b - a); };
+
+    // Lambda for getting Z test height given x input along the bresenham line.
+    auto GetZValue = [&](const int x) -> double
+    {
+        const auto xDiff = xB - xA;
+        const float xPercent = static_cast<float>(x) / xDiff;
+        return lerp(static_cast<int>(zA), static_cast<int>(zB), xPercent);
```

Comment:
why casting zA and zB to int ? Shouldn't the interpolation be done on double to be consistant. And maybe use a double for xPercent too. In general, float should be rarely used compared to using double, unless there are good reasons to do so (memory savings or processing speed on big arrays typically, which isn't hte case here)

Reply:
Oh duh, `lerp` should just be changed to take in doubles.

Reply:
Fixed!